frontend/src/main/scala/bloop/engine/State.scala

```diff
@@ -1,47 +1,78 @@
 package bloop.engine
 
 import bloop.CompilerCache
-import bloop.cli.ExitStatus
+import bloop.cli.{CommonOptions, ExitStatus}
 import bloop.engine.caches.{ResultsCache, StateCache}
 import bloop.io.Paths
-import bloop.logging.Logger
+import bloop.logging.{BloopLogger, Logger}
 
+/**
+ * Represents the state for a given build.
+ *
+ * An state is aggressively cached by bloop so that operations that target a build can reuse
+ * the state that has already been created before. The state is immutable and allows different
+ * builds to change common values utilized by all the bloop operations.
+ *
+ * @param build The build with which the state is associated.
+ * @param results The results cache that contains all the analysis file for a build.
+ * @param compilerCache The cache that has pointers to hot compilers (shared across other states).
+ * @param commonOptions The environment options necessary for input and output.
+ * @param status The status in which the state is currently.
+ * @param logger The logger that is used and is associated with a given build.
+ */
 final case class State private (
     build: Build,
     results: ResultsCache,
     compilerCache: CompilerCache,
+    commonOptions: CommonOptions,
     status: ExitStatus,
     logger: Logger
 ) {
   private[bloop] val executionContext: scala.concurrent.ExecutionContext =
     ExecutionContext.threadPool
+  /* TODO: Improve the handling and merging of different status. Use the status to report errors. */
   def mergeStatus(newStatus: ExitStatus): State =
     this.copy(status = ExitStatus.merge(status, newStatus))
 }
 
 object State {
   private[bloop] val stateCache: StateCache = StateCache.empty
-  private[bloop] val compilerCache: CompilerCache = {
-    import sbt.internal.inc.bloop.ZincInternals
-    val provider = ZincInternals.getComponentProvider(Paths.getCacheDirectory("components"))
-    new CompilerCache(provider, Paths.getCacheDirectory("scala-jars"), Logger.get)
+  private var singleCompilerCache: CompilerCache = null
+  private def getCompilerCache(logger: Logger): CompilerCache = synchronized {
+    if (singleCompilerCache != null) singleCompilerCache
+    else {
+      import sbt.internal.inc.bloop.ZincInternals
+      val provider = ZincInternals.getComponentProvider(Paths.getCacheDirectory("components"))
+      val jars = Paths.getCacheDirectory("scala-jars")
+      singleCompilerCache = new CompilerCache(provider, jars, logger)
+      singleCompilerCache
+    }
```

Comment:
Yes, this is a tricky part, the compiler cache is not owned by a build in particular, but we still need a logger. As I wanted the first build to trigger the compilation to register the logs of the compiler cache, I did it this way. I agree it's not ideal, but the other solution is to create a global bloop logger that we pass around.
  
However, this will complicate things for not much in return and has some design issues. For example, for every compile command that the server receive we need to swap the output streams. This can cause race conditions and misbehaviours.